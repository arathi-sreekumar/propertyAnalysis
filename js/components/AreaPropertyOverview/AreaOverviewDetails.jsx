// @flow

import React from 'react';
import { string, shape } from 'prop-types';

type FlowAreaData = {
	area: {
		area_name?: string,
		country?: string,
		county?: string,
		postcode?: string,
		street?: string,
		town?: string,
		zed_index?: string
	}
};

/*
  * AreaOverviewDetails component displays basic details regarding an area
*/
const AreaOverviewDetails = (props: FlowAreaData) => {
	const county = props.area.county;
	const postcode = props.area.postcode;
	const street = props.area.street;
	const town = props.area.town;
	const country = props.area.country;
	const zedIndex = props.area.zed_index;
	const areaName = props.area.area_name;

	return (
		<ul>
			{areaName && <li>Area: {areaName}</li>}
			<li>Country: {country}</li>
			<li>County: {county}</li>
			{postcode && <li>PostCode: {postcode}</li>}
			{street && <li>Street: {street}</li>}
			{town && <li>Town: {town}</li>}
			<li> Average property price: Â£{zedIndex} </li>
		</ul>
	);
};

AreaOverviewDetails.propTypes = {
	area: shape({
		area_name: string,
		country: string,
		county: string,
		postcode: string,
		street: string,
		town: string,
		zed_index: string
	}).isRequired
};

AreaOverviewDetails.defaultProps = {
	area: {
		area_name: '-',
		country: '-',
		county: '-',
		postcode: '-',
		street: '-',
		town: '-',
		zed_index: '-'
	}
};

export default AreaOverviewDetails;
