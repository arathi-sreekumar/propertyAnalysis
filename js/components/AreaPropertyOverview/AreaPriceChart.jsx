// @flow

import React from 'react';
// import Chart from 'react-d3-core';
import { LineChart } from 'react-easy-chart';
import { arrayOf, shape, string, number } from 'prop-types';

type FlowChartData = {
  data: ?({ x: string, y: number }[])
};

/*
 * AreaPriceChart component: 
 * Shows average pricing information over the last five years for selected area
*/
class AreaPriceChart extends React.Component {
  constructor(props: FlowChartData) {
    super(props);
    this.state = {
      data: props.data
    };
  }

  state: FlowChartData;

  render() {
    const width = 700;
    const height = 300;
    const chartData = [this.state.data];

    return (
      <div>
        <h2 className="subtitle">Average property price over last 5 years</h2>
        <LineChart
          xType={'text'}
          axisLabels={{ x: 'Last 5 years', y: 'Average property price' }}
          axes
          dataPoints
          width={width}
          height={height}
          interpolate={'cardinal'}
          data={chartData}
        />
      </div>
    );
  }
}

AreaPriceChart.propTypes = {
  data: arrayOf(
    shape({
      x: string,
      y: number
    })
  ).isRequired
};

export default AreaPriceChart;
