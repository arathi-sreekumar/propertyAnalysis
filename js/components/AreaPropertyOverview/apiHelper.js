import { getPropertyDetailsByAreaUrl, getPropertyDetailsByPostCodeUrl } from './requestUrls';

// function checkStatus(response) {
// 	if (response.status >= 200 && response.status < 300) {
// 		return response;
// 	}
// 	const error = new Error(response.statusText);
// 	error.response = response;
// 	throw error;
// }

export function fetchPropertyOverview(props) {
	const options = props || {};
	const { area, postcode } = options;
	let url;

	if (area) {
		url = getPropertyDetailsByAreaUrl(area);
	} else if (postcode) {
		url = getPropertyDetailsByPostCodeUrl(postcode);
	} else {
		const error = new Error('Need to provide area or postcode');
		throw error;
	}

	return fetch(url, {
		headers: {
			Accept: 'application/json',
			'Content-Type': 'application/json'
		},
		method: 'GET'
	}).then(response => {
		if (!response.ok) {
			throw Error(response.statusText);
		}
		return response.json();
	});

	// .then(res => res.json());
}

export default fetchPropertyOverview;
