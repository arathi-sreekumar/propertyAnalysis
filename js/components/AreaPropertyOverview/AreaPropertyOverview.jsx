// @flow

import React, { Component } from 'react';
import { fetchPropertyOverview } from './helpers/apiHelper';
import { extractZedIndexChartData, extractAreaDetails, getBoundingBox } from './helpers/dataHelper';
import AreaOverviewDetails from './AreaOverviewDetails';
import AreaPriceChart from './AreaPriceChart';
import GoogleMapWithMarker from '../common/GoogleMaps/MapWithAMarker';

type FlowState = {
	error: ?{
		message: string
	},
	result: ?{
		zed_index?: string,
		zed_index_3month?: number,
		zed_index_6month?: number,
		zed_index_1year?: number,
		zed_index_2year?: number,
		zed_index_3year?: number,
		zed_index_4year?: number,
		zed_index_5year?: number,
		area_name?: string,
		country?: string,
		county?: string,
		postcode?: string,
		street?: string,
		town?: string,
		zed_index?: string,
		latitude?: number,
		longitude?: number
	},
	isLoaded: boolean,
	searchTerm: ?string
};

/*
 * AreaPropertyOverview component: This creates a combined view that shows the property details,
 * average price chart and selected location on a map, after fetching the data from zoopla api.
*/
class AreaPropertyOverview extends Component {
	state: FlowState;

	state = {
		error: null,
		isLoaded: false,
		result: null,
		searchTerm: ''
	};

	componentDidMount() {
		this.fetchData();
	}

	// This is to update the results if search string changes during a rerendering
	componentDidUpdate(prevProps: { searchTerm: ?string }) {
		if (this.props.searchTerm !== prevProps.searchTerm) {
			this.setSearchState(this.props.searchTerm);

			this.fetchData();
		}
	}

	setStateAsync(state: {}) {
		return new Promise(resolve => {
			this.setState(state, resolve);
		});
	}

	setSearchState(searchTerm: ?string) {
		if (searchTerm) {
			this.setState({ searchTerm });
		}
	}

	props: {
		searchTerm: ?string
	};

	// fetchData method: Makes an api request if there is a searchTerm and handles the response
	async fetchData() {
		if (!this.props.searchTerm) {
			return;
		}

		try {
			const response = await fetchPropertyOverview(this.props.searchTerm);
			await this.setStateAsync({
				isLoaded: true,
				result: response
			});
		} catch (error) {
			this.setState({ error });
		}
	}

	render() {
		const { error, isLoaded, result } = this.state;

		if (!this.props.searchTerm) {
			return <div>You need to search for a postcode!</div>;
		}

		if (error) {
			return <div>Please enter a valid postcode! We could not find results for {this.props.searchTerm}</div>;
		}

		if (!isLoaded) {
			return <div>Loading...</div>;
		}

		if (result && Object.getOwnPropertyNames(result).length > 0) {
			const areaDetails = extractAreaDetails(result);
			const chartData = extractZedIndexChartData(result);
			const mapCenter = { lat: result.latitude, lng: result.longitude };
			const mapZoom = 11;
			const mapBounds = getBoundingBox(result.bounding_box);

			return (
				<div className="search-result">
					<AreaOverviewDetails area={areaDetails} />
					<AreaPriceChart data={chartData} />
					<GoogleMapWithMarker center={mapCenter} zoom={mapZoom} bounds={mapBounds} />
				</div>
			);
		}

		return <div>No search results found</div>;
	}
}

export default AreaPropertyOverview;
