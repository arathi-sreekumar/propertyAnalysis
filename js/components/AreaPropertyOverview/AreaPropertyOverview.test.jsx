import React from 'react';
import { shallow } from 'enzyme';
import AreaPropertyOverview from './AreaPropertyOverview';

describe('AreaPropertyOverview', () => {
	it('renders correctly', () => {
		const component = shallow(<AreaPropertyOverview />);
		expect(component).toMatchSnapshot();
	});
});
