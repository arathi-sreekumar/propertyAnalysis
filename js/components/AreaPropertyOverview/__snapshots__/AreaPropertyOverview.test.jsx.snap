// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AreaPropertyOverview renders correctly for empty search 1`] = `
<AreaPropertyOverview
  searchTerm=""
>
  <div>
    You need to search for a postcode!
  </div>
</AreaPropertyOverview>
`;

exports[`AreaPropertyOverview renders correctly for search "bn10" 1`] = `
<AreaPropertyOverview
  searchTerm="bn10"
>
  <div>
    Loading...
  </div>
</AreaPropertyOverview>
`;

exports[`AreaPropertyOverview search receiving empty data renders correctly 1`] = `
<AreaPropertyOverview
  searchTerm="bn10"
>
  <div>
    No search results found
  </div>
</AreaPropertyOverview>
`;

exports[`AreaPropertyOverview search receiving null as data renders correctly 1`] = `
<AreaPropertyOverview
  searchTerm="bn10"
>
  <div>
    No search results found
  </div>
</AreaPropertyOverview>
`;

exports[`AreaPropertyOverview search updates data correctly updates correctly 1`] = `
<AreaPropertyOverview
  searchTerm="bn11"
>
  <div
    className="search-result"
  >
    <AreaOverviewDetails
      area={
        Object {
          "area_name": undefined,
          "country": undefined,
          "county": undefined,
          "postcode": undefined,
          "street": undefined,
          "town": undefined,
          "zed_index": undefined,
        }
      }
    >
      <ul>
        <li>
          Country: 
        </li>
        <li>
          County: 
        </li>
        <li>
           Average property price: Â£
           
        </li>
      </ul>
    </AreaOverviewDetails>
    <AreaPriceChart
      data={Array []}
    >
      <div>
        <h2
          className="subtitle"
        >
          Average property price over last 5 years
        </h2>
        <LineChart
          axes={true}
          axisLabels={
            Object {
              "x": "Last 5 years",
              "y": "Average property price",
            }
          }
          clickHandler={[Function]}
          data={
            Array [
              Array [],
            ]
          }
          dataPoints={true}
          datePattern="%d-%b-%y"
          height={300}
          interpolate="cardinal"
          lineColors={Array []}
          mouseMoveHandler={[Function]}
          mouseOutHandler={[Function]}
          mouseOverHandler={[Function]}
          width={700}
          xType="text"
          yType="linear"
        >
          <div
            className="line-chart-cc84da6e5eda728c44a210ca09adbd1ad7130bac"
          >
            <Style
              rules={
                Object {
                  ".area": Object {
                    "opacity": 0.7,
                  },
                  ".area:hover": Object {
                    "opacity": 1,
                  },
                  ".axis": Object {
                    "font-family": "dobra-light,Arial,sans-serif",
                    "font-size": "9px",
                  },
                  ".axis .label": Object {
                    "font": "14px arial",
                  },
                  ".axis path, .axis line": Object {
                    "fill": "none",
                    "shape-rendering": "crispEdges",
                    "strokeWidth": 1,
                  },
                  ".bar": Object {
                    "fill": "blue",
                    "opacity": 1,
                    "transition": "x 0.35s ease-in, y 0.35s ease-in, height 0.5s ease-in, width 0.5s ease-in",
                  },
                  ".bar:hover": Object {
                    "opacity": 0.8,
                  },
                  ".dot": Object {
                    "opacity": 0.85,
                    "strokeWidth": 0,
                    "transition": "cx 0.35s ease-in, cy 0.35s ease-in, r 0.5s ease-in",
                  },
                  ".dot:hover": Object {
                    "opacity": 1,
                  },
                  ".line": Object {
                    "fill": "none",
                    "opacity": 0.7,
                    "strokeWidth": 1.5,
                  },
                  ".line:hover": Object {
                    "opacity": 1,
                  },
                  ".pie-chart-label": Object {
                    "fill": "#000",
                    "fontFamily": "sans-serif",
                    "fontSize": "12px",
                    "textAnchor": "middle",
                  },
                  ".pie-chart-slice": Object {
                    "opacity": "1",
                    "stroke": "#fff",
                    "strokeWidth": 1,
                  },
                  ".pie-chart-slice:hover": Object {
                    "opacity": "0.8",
                  },
                  ".tick line": Object {
                    "opacity": "0.7",
                    "stroke": "lightgrey",
                    "strokeWidth": 1,
                  },
                  ".x circle.tick-circle": Object {
                    "cy": "8px",
                  },
                  ".x circle.tick-circle ": Object {
                    "fill": "lightgrey",
                  },
                  ".y circle.tick-circle": Object {
                    "cx": "-8px",
                    "fill": "lightgrey",
                  },
                  ".y.axis line": Object {
                    "display": "none",
                    "stroke": "lightgrey",
                  },
                  "circle.data-point": Object {
                    "opacity": 0.7,
                    "r": 4,
                    "transition": "cx 0.35s ease-in, cy 0.35s ease-in",
                  },
                  "circle.data-point:hover": Object {
                    "opacity": 1,
                    "r": 6,
                  },
                  "circle.tick-circle": Object {
                    "fill": "lightgrey",
                    "r": 2,
                  },
                  "x.axis path": Object {
                    "display": "none",
                    "stroke": "lightgrey",
                  },
                }
              }
              scopeSelector=".line-chart-cc84da6e5eda728c44a210ca09adbd1ad7130bac"
            >
              <style
                dangerouslySetInnerHTML={
                  Object {
                    "__html": ".line-chart-cc84da6e5eda728c44a210ca09adbd1ad7130bac .pie-chart-slice{stroke: #fff;
                  stroke-width: 1;
                  opacity: 1;}.line-chart-cc84da6e5eda728c44a210ca09adbd1ad7130bac .pie-chart-slice:hover{opacity: 0.8;}.line-chart-cc84da6e5eda728c44a210ca09adbd1ad7130bac .pie-chart-label{font-family: sans-serif;
                  font-size: 12px;
                  text-anchor: middle;
                  fill: #000;}.line-chart-cc84da6e5eda728c44a210ca09adbd1ad7130bac .bar{fill: blue;
                  transition: x 0.35s ease-in, y 0.35s ease-in, height 0.5s ease-in, width 0.5s ease-in;
                  opacity: 1;}.line-chart-cc84da6e5eda728c44a210ca09adbd1ad7130bac .bar:hover{opacity: 0.8;}.line-chart-cc84da6e5eda728c44a210ca09adbd1ad7130bac .line{fill: none;
                  stroke-width: 1.5;
                  opacity: 0.7;}.line-chart-cc84da6e5eda728c44a210ca09adbd1ad7130bac .line:hover{opacity: 1;}.line-chart-cc84da6e5eda728c44a210ca09adbd1ad7130bac .area{opacity: 0.7;}.line-chart-cc84da6e5eda728c44a210ca09adbd1ad7130bac .area:hover{opacity: 1;}.line-chart-cc84da6e5eda728c44a210ca09adbd1ad7130bac .dot{stroke-width: 0;
                  opacity: 0.85;
                  transition: cx 0.35s ease-in, cy 0.35s ease-in, r 0.5s ease-in;}.line-chart-cc84da6e5eda728c44a210ca09adbd1ad7130bac .dot:hover{opacity: 1;}.line-chart-cc84da6e5eda728c44a210ca09adbd1ad7130bac circle.data-point{r: 4px;
                  opacity: 0.7;
                  transition: cx 0.35s ease-in, cy 0.35s ease-in;}.line-chart-cc84da6e5eda728c44a210ca09adbd1ad7130bac circle.data-point:hover{r: 6px;
                  opacity: 1;}.line-chart-cc84da6e5eda728c44a210ca09adbd1ad7130bac circle.tick-circle{r: 2px;
                  fill: lightgrey;}.line-chart-cc84da6e5eda728c44a210ca09adbd1ad7130bac .x circle.tick-circle{cy: 8px;}.line-chart-cc84da6e5eda728c44a210ca09adbd1ad7130bac .axis{font-family: dobra-light,Arial,sans-serif;
                  font-size: 9px;}.line-chart-cc84da6e5eda728c44a210ca09adbd1ad7130bac .axis .label{font: 14px arial;}.line-chart-cc84da6e5eda728c44a210ca09adbd1ad7130bac .axis path,.line-chart-cc84da6e5eda728c44a210ca09adbd1ad7130bac .axis line{fill: none;
                  stroke-width: 1;
                  shape-rendering: crispEdges;}.line-chart-cc84da6e5eda728c44a210ca09adbd1ad7130bac x.axis path{display: none;
                  stroke: lightgrey;}.line-chart-cc84da6e5eda728c44a210ca09adbd1ad7130bac .tick line{stroke: lightgrey;
                  stroke-width: 1;
                  opacity: 0.7;}.line-chart-cc84da6e5eda728c44a210ca09adbd1ad7130bac .x circle.tick-circle{fill: lightgrey;}.line-chart-cc84da6e5eda728c44a210ca09adbd1ad7130bac .y circle.tick-circle{cx: -8px;
                  fill: lightgrey;}.line-chart-cc84da6e5eda728c44a210ca09adbd1ad7130bac .y.axis line{display: none;
                  stroke: lightgrey;}",
                  }
                }
              />
            </Style>
            <svg
              height={300}
              style={Object {}}
              width={700}
            >
              <g
                style={Object {}}
                transform="translate(50, 20)"
              >
                <g
                  className="x axis"
                  fill="none"
                  fontFamily="sans-serif"
                  fontSize={10}
                  style={Object {}}
                  textAnchor="middle"
                  transform="translate(0, 230)"
                >
                  <path
                    className="domain"
                    d="M0.5,0V0.5H630.5V0"
                    stroke="#000"
                    style={Object {}}
                  />
                  <text
                    className="label"
                    style={
                      Object {
                        "textAnchor": "end",
                      }
                    }
                    x={630}
                    y={40}
                  >
                    Last 5 years
                  </text>
                </g>
                <g
                  className="y axis"
                  fill="none"
                  fontFamily="sans-serif"
                  fontSize={10}
                  style={Object {}}
                  textAnchor="end"
                  transform="translate(0, 0)"
                >
                  <path
                    className="domain"
                    d="M-6,230.5H0.5V0.5H-6"
                    stroke="#000"
                    style={Object {}}
                  />
                  <text
                    className="label"
                    dy=".9em"
                    style={
                      Object {
                        "textAnchor": "end",
                      }
                    }
                    transform="rotate(-90)"
                    x={0}
                    y={-50}
                  >
                    Average property price
                  </text>
                </g>
                <g
                  className="lineChart"
                  style={Object {}}
                >
                  <path
                    className="line"
                    style={
                      Object {
                        "stroke": "#3F4C55",
                      }
                    }
                  />
                </g>
                <g
                  className="dataPoints"
                  style={Object {}}
                />
              </g>
            </svg>
          </div>
        </LineChart>
      </div>
    </AreaPriceChart>
    <GoogleMapWithMarker
      bounds={null}
      center={
        Object {
          "lat": undefined,
          "lng": undefined,
        }
      }
      zoom={11}
    >
      <withScriptjs(withGoogleMap(Component))
        bounds={null}
        center={
          Object {
            "lat": undefined,
            "lng": undefined,
          }
        }
        containerElement={
          <div
            style={
              Object {
                "height": "400px",
              }
            }
          />
        }
        googleMapURL="https://maps.googleapis.com/maps/api/js?key=AIzaSyCBUAx3W_ykWX9gSePohkLo-TRTeNV9iwI&v=3.exp&libraries=geometry,drawing,places"
        loadingElement={
          <div
            style={
              Object {
                "height": "100%",
              }
            }
          />
        }
        mapElement={
          <div
            style={
              Object {
                "height": "100%",
              }
            }
          />
        }
        zoom={11}
      >
        <div
          style={
            Object {
              "height": "100%",
            }
          }
        />
      </withScriptjs(withGoogleMap(Component))>
    </GoogleMapWithMarker>
  </div>
</AreaPropertyOverview>
`;

exports[`AreaPropertyOverview search with failed api call renders correctly 1`] = `
<AreaPropertyOverview
  searchTerm="peacehaven"
>
  <div>
    Please enter a valid postcode! We could not find results for 
    peacehaven
  </div>
</AreaPropertyOverview>
`;
